src/Posts.tsx

```diff
@@ -1,50 +1,214 @@
 import React from 'react';
+import Moment from 'react-moment';
 
-const API = 'http://localhost:4000/api/v1/posts';
+const API = 'http://localhost:4000/api/v1/';
 type Props = {};
 
 type State = {
   posts: Post[];
+  categories: Category[];
   isLoading: boolean;
-  errors: Error[];
+  errors: any;
+  newPostErrors: string;
+  formIsValid: boolean;
+  topicValue: string;
+  topicValid: boolean;
+  postValue: string;
+  postValid: boolean;
+  categoryValue: string;
+  pinnedValue: boolean;
+  postsLoading: boolean;
+  categoriesLoading: boolean;
+  topicError: string;
+  postError: string;
 };
 type Post = {
   id: number;
   post: string;
   topic: string;
-  datetime: Date;
+  created: Date;
   pinned: boolean;
   modified: Date;
 };
+type Category = {
+  id: number;
+  name: string;
+  description: string;
+};
 
 class Posts extends React.Component<Props, State> {
   constructor(props: Props) {
     super(props);
     this.state = {
       posts: [],
       isLoading: false,
-      errors: []
+      errors: undefined,
+      newPostErrors: '',
+      categories: [],
+      formIsValid: false,
+      topicValue: '',
+      topicValid: false,
+      postValue: '',
+      postValid: false,
+      pinnedValue: false,
+      categoryValue: '1',
+      postsLoading: false,
+      categoriesLoading: false,
+      topicError: '',
+      postError: ''
     };
   }
   componentDidMount() {
-    this.getAndSetPosts();
+    this.loadPostsAndCategories();
   }
+  loadPostsAndCategories = async () => {
+    this.setState({ isLoading: true, errors: undefined });
+    await this.getAndSetPosts();
+    await this.getCategories();
+  };
   getAndSetPosts = async () => {
-    this.setState({ isLoading: true, errors: [] });
-    fetch(API)
+    this.setState({ postsLoading: true });
+    fetch(API + 'posts')
+      .then(response => {
+        if (response.status === 404) {
+          return [];
+        } else if (response.ok) {
+          return response.json();
+        } else {
+          throw Error(response.statusText);
+        }
+      })
+      .then(data => this.setState({ posts: data, postsLoading: false }))
```

Comment:
Maybe sort the list here so that the latest posted post is the first on the list? People are interested on the latest stuff right ;)